ast/src/analyzed/types.rs

```diff
@@ -116,25 +198,86 @@ pub struct FunctionType {
     pub value: Box<Type>,
 }
 
-impl FunctionType {
-    /// Returns the column type `int -> fe`.
-    pub fn col() -> Self {
+impl<T: FieldElement, Ref: Display> From<FunctionTypeName<Expression<T, Ref>>> for FunctionType {
+    fn from(name: FunctionTypeName<Expression<T, Ref>>) -> Self {
         FunctionType {
-            params: vec![Type::Int],
-            value: Box::new(Type::Fe),
+            params: name.params.into_iter().map(Into::into).collect(),
+            value: Box::new(Type::from(*name.value)),
         }
     }
+}
 
-    pub fn needs_parentheses(&self) -> bool {
-        *self != Self::col()
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize, JsonSchema)]
+pub struct TypeScheme {
+    /// Type variables and their trait bounds
+    pub vars: TypeBounds,
+    /// The actual type (using the type variables from `vars`)
+    pub ty: Type,
+}
+
+impl TypeScheme {
+    /// Returns a new type scheme with type variables renamed to `T1`, `T2`, ...
+    /// (or just `T` if it is a single type variable).
+    pub fn simplify_type_vars(self) -> TypeScheme {
+        let name_substitutions: HashMap<_, _> = match self.vars.len() {
+            0 => return self,
+            1 => {
+                let var = self.vars.vars().next().unwrap();
+                [(var.clone(), "T".to_string())].into()
+            }
+            _ => self
+                .vars
+                .vars()
+                .enumerate()
+                .map(|(i, v)| ((*v).clone(), format!("T{}", i + 1)))
```

Comment:
why do we need these concrete names?

Reply:
This is used for error reporting or if two type schemes are the same up to renaming of type variables.